### DISCUSSION TITLE
building c bindings for gpufaiss

### DISCUSSION THREAD

**Discussion by orihabAnyvision** [facebookresearch/faiss#2000] [Q&A] [answered]
I must be missing something dumb but I cant seem to build it. 

I built faiss with DFAISS_ENABLE_C_API=ON and I get under the build folder the c_api folder, which has libfaiss_c.so. It has as well a gpu folder, which in the install [here](https://github.com/facebookresearch/faiss/blob/master/c_api/INSTALL.md) is says just cd into the gpu folder and do make. But that doesn't seem to do anything, and I dont see any other target in the makefile which seems relevant.. 

What am I missing?

**Comment by Enet4** [comment, marked as answer]
Oh, that part of the documentation is out of date. It used to work that way, but the process is now fully integrated into CMake since version 1.7.1. The `faiss_c` library will also expose the GPU part of the C API when built with the right options.

**Reply by orihabAnyvision** [reply]
ok. so the c bindings for the gpu methods are inside libfaiss_c.so? Sorry I'm new to c bindings and am trying to wrap them with go code...

**Reply by Enet4** [reply]
Yes, that would be the case. But if the purpose is to use Faiss from Go, I believe that such bindings already exist: https://github.com/41tair/go-faiss and https://github.com/DataIntelligenceCrew/go-faiss

**Reply by orihabAnyvision** [reply]
yes I know those 2 repos.. the 41tair only implemented a small portion of the functionality.. The dataintelligenceCrew did a bit more, I actually opened a PR earlier which added the index_idMap functionality, but they also didn't bind the GPU functionality so I'm looking to add that bit as well..

**Comment by uhbuhb** [comment]
Hey @Enet4, in continuation to the previous.. I realized that the releaseGpuResource isn't implemented as well, and am having trouble implementing it.. 

Im getting the 2 errors which hint to me that I'm close but missing the cast functionality:

```
cannot use _cgo1 (type **_Ctype_struct_FaissGpuResourcesProvider_H) as type **_Ctype_struct_FaissGpuResources_H in argument to _Cfunc_faiss_GpuResourcesProvider_getResources 
cannot use _cgo0 (type *_Ctype_struct_FaissGpuResourcesProvider_H) as type *_Ctype_struct_FaissGpuResources_H in argument to _Cfunc_faiss_GpuResources_free
```

I see in StandardGpuResource_c.h the classes are defined with FAISS_DECLARE_CLASS_INHERITED(StandardGpuResources, GpuResourcesProvider)..
I see in GpuResources_c.cpp there is the following line of code:

reinterpret_cast<GpuResourcesProvider*>(res)->getResources()
which I dont really understand (I dont know cpp), but from the error I'm getting it seems like I have everything in order except for the cast part.

Any insight as to how I could call the reinterpret_cast functionality from go?

**Reply by Enet4** [reply]
So, here is how one would create and release GPU resources in C (example taken from [here](https://github.com/facebookresearch/faiss/blob/main/c_api/gpu/example_gpu_c.c)).

```c
    FaissStandardGpuResources* gpu_res = NULL;
    int err_code = faiss_StandardGpuResources_new(&gpu_res);

    // use gpu_res

    faiss_GpuResources_free(gpu_res);
```

The important part here is that no casting was required because the type declarations are already aliased to the same underlying type: `FaissStandardGpuResources` and `FaissGpuResourcesProvider` are the same opaque type underneath. Apparently, this used to be the case for `FaissGpuResources` as well, but some recent changes might have changed that.

[`reinterpret_cast`](https://en.cppreference.com/w/cpp/language/reinterpret_cast) is just a way to perceive a pointer type as another pointer type, so that the types declared in the C API are opaque. It is usually a no-op. You should be able to get the same behavior of `reinterpret_cast` by just casting one pointer type to the other, thus mitigating the issues seen here.

**Reply by uhbuhb** [reply]
Hi @Enet4  thanks for the reply! I saw the example and am trying to copy after it, but still Im getting a casting error. This I think is the most boiled down version of a function which does only what you suggested, create a gpuResource and free it:

```
func CreateAndFree() error {
	var gpuResource *C.FaissStandardGpuResources
	c := C.faiss_StandardGpuResources_new(&gpuResource)
	if c != 0 {
		return errors.New("error on init gpu")
	}
	C.faiss_GpuResources_free(gpuResource)
	return nil
}
```

But still on compiling I get the casting error:
```
vendor/github.com/DataIntelligenceCrew/go-faiss/gpu_bindings.go:72:101: cannot use _cgo0 (type *_Ctype_struct_FaissGpuResourcesProvider_H) as type *_Ctype_struct_FaissGpuResources_H in argument to _Cfunc_faiss_GpuResources_free
```

I saw theres a function faiss_GpuResourcesProvider_getResources in GpuResources_c.cpp - it hasnt helped me but maybe you would have some insight  - any chance this could be helpful in getting a resource from a resourceProvidor?

**Reply by uhbuhb** [reply]
ok FYI anyone who finds this thread - the answer is to use faiss_**Standard**GpuResources_free and not faiss_GpuResources_free!! @Enet4 would be useful if this was updated in the docs/example as I spent a bunch of time trying to figure out :facepalm:

**Reply by Enet4** [reply]
@uhbuhb Yup, some tweaking may be needed here. In practice, `faiss_GpuResources_free` should have worked just the same, but as I said earlier, the alias between `GpuResources` and `StandardGpuResources` was lost. And if these problems still emerge, the type system can be circumvented by casting the pointer.
